### PR TITLE
Visualizer

### PR DESCRIPTION
**Pull request by evouga**
Fixes https://github.com/Kattis/problem-package-format/issues/138

### CODE REVIEW

File: spec/2023-07-draft.md [hunk 1]
```diff
@@ -801,6 +801,10 @@ A judging system that implements this format must support the `judgemessage.txt`
 (I.e., content of the `judgemessage.txt` file, if produced by the validator, must be provided by the judging system to a human judge examining the submission).
 Having the judging system support other files is optional.
 
+The validator may create one or more image files in the feedback directory with name `teamimage.ext` and/or `judgeimage.ext`, where `ext` is an appropriate image file format extension.
```
Comment (pehrsoderman) [inline spec/2023-07-draft.md:804]: What are the allowed extensions?

**Comment by RagnarGrootKoerkamp** [comment]
@mzuenni @mpsijm see also the last message in the referenced issue. This is now speccing output visualisers, but probably we should also keep input visualisers in some way

File: spec/2023-07-draft.md (hunk 1 above)
Comment (evouga) [reply spec/2023-07-draft.md:804]: Do we need to specify, if we are not requiring judge systems to support any particular formats?

Comment (niemela) [reply spec/2023-07-draft.md:808]: We specify supported image file formats in many places. I think it makes sense to do so here as well.

Currently we allow:
- `.png`, `.jpg`/`.jpeg`, `.pdf` for problem statement images.
- `.png`, `.jpg`/`.jpeg`, `.svg` for annotations.

Output visualisers images are basically the equivalent for "annotations", but for outputs (and nit inputs), so I think it makes sense to allow the exact same set of formats.

**Review by niemela — CHANGES_REQUESTED** [review]
Specify allowed image formats (`.png`, `.jpg`/`.jpeg`, `.svg`)

**Review by niemela — APPROVED** [review]

**Review by niemela — APPROVED** [review]

**Comment by Tagl** [comment]
> JPEG images group together the following filename extensions:
> 
>     .jpg
>     .jpeg
>     .jpe
>     .jif
>     .jfif
>     .jfi

Of course it isn't easy

**Comment by eldering** [comment]
> > JPEG images group together the following filename extensions:
> > ```
> > .jpg
> > .jpeg
> > .jpe
> > .jif
> > .jfif
> > .jfi
> > ```
> 
> Of course it isn't easy

I agree with Fredrik that just `.jpg` and `.jpeg` are needed.

**Review by Tagl — APPROVED** [review]

File: spec/2023-07-draft.md [hunk 2]
```diff
@@ -848,6 +856,20 @@ Almost all test cases failed — are you even trying to solve the problem?
 A validator program is allowed to write any kind of debug information to its standard error pipe.
 This information may be displayed to the user upon invocation of the validator.
 
+### Output Visualizer
+
+An output visualizer is an optional [program](#programs) that is run after every invocation of the output validator in order to generate images illustrating the submission output. 
+A visualizer program must be an application (executable or interpreted) capable of being invoked with a command line call. It is invoked using the same arguments as the output validator. 
+It must be provided as the directory `output_visualizer/`.
+It must either be written in C++ or Python 3,
```
Comment (RagnarGrootKoerkamp) [inline spec/2023-07-draft.md:864]: Looking over recent problems, we have used
- `asymptote`, some very specific graphic drawing language
- `matplotlib`
- a custom c++ lib that writes svg files
- a custom python function to write pngs.

Restricting to python/c++ by default sgtm, although these may want to depend on libs that need to be installed by the system.

File: spec/2023-07-draft.md [hunk 3]
```diff
@@ -801,6 +805,10 @@ A judging system that implements this format must support the `judgemessage.txt`
 (I.e., content of the `judgemessage.txt` file, if produced by the validator, must be provided by the judging system to a human judge examining the submission).
 Having the judging system support other files is optional.
 
+The validator may create one or more image files in the feedback directory with name `teamimage.ext` and/or `judgeimage.ext`, where `ext` is one of: `png`, `jpg`, `jpeg`, or `svg`.
```
Comment (RagnarGrootKoerkamp) [inline spec/2023-07-draft.md:808]: Also say that the output validator may create additional metadata files that can help the visualizer

File: spec/2023-07-draft.md (hunk 2 above)
Comment (evouga) [reply spec/2023-07-draft.md:864]: Should we say something about this in the spec? (Do we want to require matplotlib?)

**Review by RagnarGrootKoerkamp — APPROVED** [review]

**Comment by RagnarGrootKoerkamp** [comment]
We also discussed allowing passing additional flags to the visualizer, e.g. to disable visualising large cases with number of nodes `>= n`. Then that could also be changed by a contest system during the contest.

**Comment by evouga** [comment]
What is "number of nodes"?

Can the judge system just ignore/delete image files it deems too large? Or is the idea to save on CPU cost of judging the submission?

**Comment by RagnarGrootKoerkamp** [comment]
As in, for example the number of nodes in an input graph. Indeed, the idea is to not spend time rendering cases that are way too large so that they take very long to render anyway and probably wouldn't be helpful to inspect anyway.

**Pull request merged**